tools/flaky-test-reporter/slack_notification.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// slack_notification.go sends notifications to slack channels
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strings"
+	"sync"
+	"time"
+
+	"net/http"
+	"net/url"
+
+	"github.com/knative/test-infra/shared/testgrid"
+)
+
+const (
+	knativeBotName          = "Knative Testgrid Robot"
+	slackChatPostMessageURL = "https://slack.com/api/chat.postMessage"
+	// default filter for testgrid link
+	testgridFilter = "exclude-non-failed-tests=20"
+)
+
+var (
+	// jobNameTestgridURLMap contains harded coded mapping of job name: Testgrid tab URL relative to base URL
+	jobNameTestgridURLMap = map[string]string{
+		"ci-knative-serving-continuous": "knative-serving#continuous",
+	}
+	// slackChannelsMap defines mapping of repo: slack channels
+	slackChannelsMap = map[string][]slackChannel{
+		"serving": []slackChannel{{"test", "CA1DTGZ2N"}},
+	}
+)
+
+// SlackClient contains Slack bot related information
+type SlackClient struct {
+	userName  string
+	tokenStr  string
+	iconEmoji *string
+}
+
+// slackChannel contains channel logical name and Slack identity
+type slackChannel struct {
+	name, identity string
+}
+
+// newSlackClient reads token file and stores it for later authentication
+func newSlackClient(slackTokenPath string) (*SlackClient, error) {
+	b, err := ioutil.ReadFile(slackTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	return &SlackClient{
+		userName: knativeBotName,
+		tokenStr: strings.TrimSpace(string(b)),
+	}, nil
+}
+
+// postMessage does http post
+func (c *SlackClient) postMessage(url string, uv *url.Values) error {
+	resp, err := http.PostForm(url, *uv)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	t, _ := ioutil.ReadAll(resp.Body)
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("http response code is not 200 '%s'", string(t))
+	}
+	// response code could also be 200 if channel doesn't exist, parse response body to find out
+	var b struct {
+		OK bool `json:"ok"`
+	}
+	if err = json.Unmarshal(t, &b); nil != err || !b.OK {
+		return fmt.Errorf("response not ok '%s'", string(t))
+	}
+	return nil
+}
+
+// writeSlackMessage posts text to channel
+func (c *SlackClient) writeSlackMessage(text, channel string) error {
+	uv := &url.Values{}
+	uv.Add("username", c.userName)
+	uv.Add("token", c.tokenStr)
+	if nil != c.iconEmoji {
```

Comment:
we can remove this condition and just add this directly.
uv := url.Value {
username: c.userName,
token: c.tokenStr,
icon_emoji: c.iconEmoji,
...
}

return c.postMessage(slackUrl, &uv)

Reply:
the structure of `uv` is `map[string][]string`, so it would be more like `username: append(uv["username"], c.userName)`. So `.Add` seems like a better option